v2/internal/reconcilers/arm/azure_generic_arm_reconciler_instance.go

```diff
@@ -248,10 +251,59 @@ func (r *azureDeploymentReconcilerInstance) MonitorDelete(ctx context.Context) (
 	return ctrl.Result{Requeue: true, RequeueAfter: retryAfter}, nil
 }
 
-func (r *azureDeploymentReconcilerInstance) BeginCreateOrUpdateResource(ctx context.Context) (ctrl.Result, error) {
+func (r *azureDeploymentReconcilerInstance) BeginCreateOrUpdateResource(
+	ctx context.Context,
+) (ctrl.Result, error) {
 	if r.Obj.AzureName() == "" {
 		err := errors.New("AzureName was not set. A webhook should default this to .metadata.name if it was omitted. Is the ASO webhook service running?")
-		return ctrl.Result{}, conditions.NewReadyConditionImpactingError(err, conditions.ConditionSeverityError, conditions.ReasonFailed)
+		return ctrl.Result{},
+			conditions.NewReadyConditionImpactingError(
+				err, conditions.ConditionSeverityError, conditions.ReasonFailed)
+	}
+
+	// Create a checker for access to the extension point, if required
+	// (We always get a checker back, even if it does nothing, so we don't need to check for nil)
+	checker, refreshRequired := extensions.CreatePreReconciliationChecker(r.Extension, r.alwaysReconcile)
+	if refreshRequired {
+		// Checker requires our resource to have an up-to-date status
+		r.Log.V(Verbose).Info("Refreshing Status of resource")
+		err := r.updateStatus(ctx)
+		if err != nil && !genericarmclient.IsNotFoundError(err) {
+			// We have an error, and it's not because the resource doesn't exist yet
+			// We assume the error isn't fatal, and we'll try again later
+			return ctrl.Result{},
+				conditions.NewReadyConditionImpactingError(
+					err, conditions.ConditionSeverityWarning, conditions.ReasonFailed)
+		}
+	}
+
+	// Run our pre-reconcilation checker
+	check, err := checker(ctx, r.Obj, r.KubeClient, r.ARMClient, r.Log)
+	if err != nil {
+		// Something went wrong running the check. We assume it's not fatal, and we'll try again later
+		// Make sure we return a ReadyConditionImpactingError so that the Ready condition is updated for the user
+		// Ideally any implementation of the checker should return a ReadyConditionImpactingError, but we can't
+		// guarantee that, so we wrap as required
+		impactingError, ok := conditions.AsReadyConditionImpactingError(err)
+		if !ok {
+			impactingError = conditions.NewReadyConditionImpactingError(
+				err,
+				conditions.ConditionSeverityWarning,
+				conditions.ReasonAzureResourceNotReady)
+		}
+
+		return ctrl.Result{}, impactingError
+	}
+
+	// If the check says we don't need to reconcile, we're done
+	// We use a ReadyConditionImpactingError here to ensure the Ready condition is updated so the user can see why
+	// we're not reconciling right now. We'll try again later.
+	if !check.ShouldReconcile() {
+		return ctrl.Result{},
```

Comment:
Aren't there two cases?
1. The resource doesn't need to reconcile because it _can't yet_ (for example, `ProvisioningState` is updating currently so no point in sending a request?). This is definitely a `Warning` as you've set up here.
2. The resource doesn't need to reconcile because it's already in a good state (i.e. we wrote custom diffing logic in this handler and everything is OK.

How do we represent that second case here? That feels like it should be `Condition` == `Ready` (i.e. no `ReadyConditionImpactingError`, we're just done).

Reply:
This maybe gets at the contract w/ this extension, is `err` only for exceptional cases, or can the extension use `err` to say: "The resource is in an -ing state, so we can't reconcile now"? In which case, the upper `if` covers case 1 in my question above and this lower case could cover case 2, either by assuming you're in that case or checking something on `check`

Reply:
This is why the result has a `ShouldReconcile()` method, so we can extend the type in the future.

Reply:
We now have three factory methods for the result struct, which I think gives us the semantics we want.